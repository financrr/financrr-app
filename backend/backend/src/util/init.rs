use std::fmt::Debug;
use std::process;

use tracing::error;

pub(crate) fn expect_or_exit<T, E: Debug>(result: Result<T, E>, message: &str) -> T {
    match result {
        Ok(value) => value,
        Err(err) => {
            error!("{}: {:?}", message, err);
            process::exit(1);
        }
    }
}
