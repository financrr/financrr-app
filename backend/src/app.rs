use crate::initializers::openapi::OpenApiInitializer;
use crate::initializers::services::ServicesInitializer;
use crate::utils::folder::{create_necessary_folders, STORAGE_FOLDER};
use crate::{controllers, models::_entities::users, tasks, workers::downloader::DownloadWorker};
use async_trait::async_trait;
use loco_rs::cache::Cache;
use loco_rs::storage::Storage;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::{BackgroundWorker, Queue},
    boot::{create_app, BootResult, StartMode},
    cache,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    storage,
    task::Tasks,
    Result,
};
use migration::Migrator;
use mimalloc::MiMalloc;
use sea_orm::DatabaseConnection;
use std::path::Path;

#[global_allocator]
static GLOBAL: MiMalloc = MiMalloc;

pub struct App;

#[async_trait]
impl Hooks for App {
    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA").or(option_env!("GITHUB_SHA")).unwrap_or("dev")
        )
    }

    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    async fn boot(mode: StartMode, environment: &Environment) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![Box::new(OpenApiInitializer), Box::new(ServicesInitializer)])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .prefix("/api/v1")
            .add_route(controllers::auth::routes())
            .add_route(controllers::session::routes())
            .add_route(controllers::openapi::routes())
    }

    async fn after_context(ctx: AppContext) -> Result<AppContext> {
        create_necessary_folders()?;

        Ok(AppContext {
            cache: Cache::new(cache::drivers::inmem::new()).into(),
            storage: Storage::single(storage::drivers::local::new_with_prefix(STORAGE_FOLDER)?).into(),
            ..ctx
        })
    }

    async fn connect_workers(ctx: &AppContext, queue: &Queue) -> Result<()> {
        queue.register(DownloadWorker::build(ctx)).await?;
        Ok(())
    }
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::seed::SeedData);
        // tasks-inject (do not remove)
    }

    async fn truncate(db: &DatabaseConnection) -> Result<()> {
        // TODO add all other tables
        truncate_table(db, users::Entity).await?;
        Ok(())
    }

    async fn seed(db: &DatabaseConnection, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(db, &base.join("users.yaml").display().to_string()).await?;
        Ok(())
    }
}
