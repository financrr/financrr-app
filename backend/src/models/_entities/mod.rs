//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.10

pub mod prelude;

pub mod bank_accounts;
pub mod categories;
pub mod currencies;
pub mod external_bank_institutions;
pub mod file_attachments;
pub mod fixtures;
pub mod go_cardless_enduser_agreements;
pub mod go_cardless_requisitions;
pub mod imported_bank_accounts;
pub mod instances;
pub mod opensearch_migrations;
pub mod pending_transactions;
pub mod recurring_transactions;
pub mod sessions;
pub mod taggings;
pub mod tags;
pub mod transaction_parties;
pub mod transaction_templates;
pub mod transactions;
pub mod user_permissions;
pub mod users;
