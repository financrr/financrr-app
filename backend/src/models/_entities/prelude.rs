//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.1

pub use super::bank_accounts::Entity as BankAccounts;
pub use super::budget_criteria::Entity as BudgetCriteria;
pub use super::budget_criteria_bank_accounts::Entity as BudgetCriteriaBankAccounts;
pub use super::budget_criteria_categories::Entity as BudgetCriteriaCategories;
pub use super::budget_criteria_external_bank_accounts::Entity as BudgetCriteriaExternalBankAccounts;
pub use super::budget_criteria_tags::Entity as BudgetCriteriaTags;
pub use super::budget_histories::Entity as BudgetHistories;
pub use super::budgets::Entity as Budgets;
pub use super::categories::Entity as Categories;
pub use super::contracts::Entity as Contracts;
pub use super::currencies::Entity as Currencies;
pub use super::external_bank_account_ibans::Entity as ExternalBankAccountIbans;
pub use super::external_bank_accounts::Entity as ExternalBankAccounts;
pub use super::external_bank_institutions::Entity as ExternalBankInstitutions;
pub use super::file_attachments::Entity as FileAttachments;
pub use super::inactive_contracts::Entity as InactiveContracts;
pub use super::instances::Entity as Instances;
pub use super::linked_back_accounts::Entity as LinkedBackAccounts;
pub use super::pending_transactions::Entity as PendingTransactions;
pub use super::recurring_transactions::Entity as RecurringTransactions;
pub use super::sessions::Entity as Sessions;
pub use super::taggings::Entity as Taggings;
pub use super::tags::Entity as Tags;
pub use super::transaction_parties::Entity as TransactionParties;
pub use super::transaction_templates::Entity as TransactionTemplates;
pub use super::transactions::Entity as Transactions;
pub use super::user_permissions::Entity as UserPermissions;
pub use super::users::Entity as Users;
