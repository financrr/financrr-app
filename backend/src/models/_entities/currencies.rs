//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.4

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "currencies")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i64,
    pub user_id: Option<i64>,
    #[sea_orm(column_type = "Text")]
    pub name: String,
    #[sea_orm(column_type = "Text")]
    pub symbol: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub iso_code: Option<String>,
    pub decimal_places: i32,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_one = "super::bank_accounts::Entity")]
    BankAccounts,
    #[sea_orm(has_many = "super::pending_transactions::Entity")]
    PendingTransactions,
    #[sea_orm(has_many = "super::recurring_transactions::Entity")]
    RecurringTransactions,
    #[sea_orm(has_many = "super::transaction_templates::Entity")]
    TransactionTemplates,
    #[sea_orm(has_many = "super::transactions::Entity")]
    Transactions,
    #[sea_orm(
        belongs_to = "super::users::Entity",
        from = "Column::UserId",
        to = "super::users::Column::Id",
        on_update = "NoAction",
        on_delete = "Cascade"
    )]
    Users,
}

impl Related<super::bank_accounts::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::BankAccounts.def()
    }
}

impl Related<super::pending_transactions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::PendingTransactions.def()
    }
}

impl Related<super::recurring_transactions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::RecurringTransactions.def()
    }
}

impl Related<super::transaction_templates::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::TransactionTemplates.def()
    }
}

impl Related<super::transactions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Transactions.def()
    }
}

impl Related<super::users::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Users.def()
    }
}
