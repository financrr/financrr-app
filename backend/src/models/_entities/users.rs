//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.3

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "users")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i64,
    #[sea_orm(column_type = "Text", unique)]
    pub email: String,
    #[sea_orm(column_type = "Text")]
    pub name: String,
    pub flags: i32,
    #[sea_orm(column_type = "Text")]
    pub password: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub reset_token: Option<String>,
    pub reset_sent_at: Option<DateTimeWithTimeZone>,
    #[sea_orm(column_type = "Text", nullable)]
    pub email_verification_token: Option<String>,
    pub email_verification_sent_at: Option<DateTimeWithTimeZone>,
    pub email_verified_at: Option<DateTimeWithTimeZone>,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::categories::Entity")]
    Categories,
    #[sea_orm(has_many = "super::currencies::Entity")]
    Currencies,
    #[sea_orm(has_many = "super::sessions::Entity")]
    Sessions,
    #[sea_orm(has_many = "super::tags::Entity")]
    Tags,
    #[sea_orm(has_many = "super::user_permissions::Entity")]
    UserPermissions,
}

impl Related<super::categories::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Categories.def()
    }
}

impl Related<super::currencies::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Currencies.def()
    }
}

impl Related<super::sessions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Sessions.def()
    }
}

impl Related<super::tags::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Tags.def()
    }
}

impl Related<super::user_permissions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::UserPermissions.def()
    }
}
