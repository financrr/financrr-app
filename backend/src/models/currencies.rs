use super::_entities::currencies::{ActiveModel, Column, Entity, Model};
use crate::error::app_error::{AppError, AppResult};
use sea_orm::entity::prelude::*;

pub type Currencies = Entity;

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    // extend activemodel below (keep comment for generators)

    async fn before_save<C>(mut self, _db: &C, insert: bool) -> loco_rs::Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        if !self.updated_at.is_set() {
            self.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        if insert {
            self.created_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        Ok(self)
    }
}

impl Model {
    pub const UNKNOWN_CURRENCY_ISO_CODE: &str = "XXX";

    pub async fn get_default_currency(db: &DbConn) -> AppResult<Self> {
        Entity::find()
            .filter(Column::IsoCode.eq(Self::UNKNOWN_CURRENCY_ISO_CODE.to_string()))
            .filter(Column::UserId.is_null())
            .one(db)
            .await?
            .ok_or_else(AppError::EntityNotFound)
    }
}
