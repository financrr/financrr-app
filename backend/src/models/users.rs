pub use super::_entities::users::{self, ActiveModel, Entity, Model};
use crate::controllers::user::RegisterParams;
use crate::error::app_error::{AppError, AppResult};
use crate::models::_entities::sessions;
use crate::services::secret_generator::SecretGenerator;
use crate::services::snowflake_generator::SnowflakeGenerator;
use chrono::offset::Local;
use enumflags2::_internal::RawBitFlags;
use enumflags2::bitflags;
use loco_rs::{hash, prelude::*};
use sea_orm::prelude::Expr;
use sea_orm::sea_query::IntoCondition;
use sea_orm::{JoinType, PaginatorTrait, QuerySelect, RelationTrait};

#[bitflags(default = User)]
#[repr(u8)]
#[derive(Copy, Clone, Debug, PartialEq)]
pub enum UserFlags {
    Admin = 0b0001,
    User = 0b0010,
}

#[async_trait::async_trait]
impl ActiveModelBehavior for super::_entities::users::ActiveModel {
    async fn before_save<C>(mut self, _db: &C, insert: bool) -> Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        if !self.updated_at.is_set() {
            self.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        if insert {
            self.created_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        Ok(self)
    }
}

impl super::_entities::users::Model {
    /// finds a user by the provided email
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_email(db: &DatabaseConnection, email: &str) -> AppResult<Option<Self>> {
        let user = users::Entity::find()
            .filter(query::condition().eq(users::Column::Email, email).build())
            .one(db)
            .await?;

        Ok(user)
    }

    pub async fn is_email_unique(db: &DatabaseConnection, email: &str) -> AppResult<bool> {
        let user = users::Entity::find()
            .select_only()
            .column(users::Column::Email)
            .filter(users::Column::Email.eq(email))
            .count(db)
            .await?;

        Ok(user == 0)
    }

    /// finds a user by the provided verification token
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_verification_token(
        db: &DatabaseConnection,
        email: &str,
        token: &str,
    ) -> Result<Option<Self>, AppError> {
        Ok(users::Entity::find()
            .filter(
                query::condition()
                    .eq(users::Column::Email, email)
                    .eq(users::Column::EmailVerificationToken, token)
                    .build(),
            )
            .one(db)
            .await?)
    }

    /// finds a user by the provided reset token
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_reset_token(db: &DatabaseConnection, email: &str, token: &str) -> ModelResult<Self> {
        let user = users::Entity::find()
            .filter(
                query::condition()
                    .eq(users::Column::Email, email)
                    .eq(users::Column::ResetToken, token)
                    .build(),
            )
            .one(db)
            .await?;
        user.ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_id(db: &DatabaseConnection, id: i64) -> ModelResult<Self> {
        let user = users::Entity::find()
            .filter(query::condition().eq(users::Column::Id, id).build())
            .one(db)
            .await?;
        user.ok_or_else(|| ModelError::EntityNotFound)
    }

    /// finds a user by the provided api key
    ///
    /// # Errors
    ///
    /// When could not find user by the given token or DB query error
    pub async fn find_by_api_key(db: &DatabaseConnection, api_key: &str) -> ModelResult<Self> {
        let cloned_api_key = api_key.to_string();
        let user = users::Entity::find()
            .join(
                JoinType::LeftJoin,
                sessions::Relation::Users.def().rev().on_condition(move |_left, right| {
                    Expr::col((right, sessions::Column::ApiKey))
                        .eq(cloned_api_key.as_str())
                        .into_condition()
                }),
            )
            .one(db)
            .await?;
        user.ok_or_else(|| ModelError::EntityNotFound)
    }

    /// Verifies whether the provided plain password matches the hashed password
    ///
    /// # Errors
    ///
    /// when could not verify password
    #[must_use]
    pub fn verify_password(&self, password: &str) -> bool {
        hash::verify_password(password, &self.password)
    }

    /// Asynchronously creates a user with a password and saves it to the
    /// database.
    ///
    /// # Errors
    ///
    /// When could not save the user into the DB
    pub async fn create_with_password(
        db: &DatabaseConnection,
        snowflake_generator: &SnowflakeGenerator,
        params: &RegisterParams,
    ) -> Result<Self, AppError> {
        let txn = db.begin().await?;

        let password_hash = hash::hash_password(&params.password)?;
        let user = users::ActiveModel {
            id: ActiveValue::set(snowflake_generator.next_id().map_err(|e| ModelError::Any(e.into()))?),
            email: ActiveValue::set(params.email.to_string()),
            password: ActiveValue::set(password_hash),
            reset_token: Default::default(),
            reset_sent_at: Default::default(),
            email_verification_token: Default::default(),
            email_verification_sent_at: Default::default(),
            email_verified_at: Default::default(),
            name: ActiveValue::set(params.name.to_string()),
            flags: ActiveValue::set(UserFlags::DEFAULT as i32),
            created_at: Default::default(),
            updated_at: Default::default(),
        }
        .insert(&txn)
        .await?;

        txn.commit().await?;

        Ok(user)
    }
}

impl super::_entities::users::ActiveModel {
    /// Sets the email verification information for the user and
    /// updates it in the database.
    ///
    /// This method is used to record the timestamp when the email verification
    /// was sent and generate a unique verification token for the user.
    ///
    /// # Errors
    ///
    /// when has DB query error
    pub async fn set_email_verification_sent(
        mut self,
        db: &DatabaseConnection,
        secret_generator: &SecretGenerator,
    ) -> ModelResult<Model> {
        const EMAIL_VERIFICATION_TOKEN_LENGTH: usize = 8;
        self.email_verification_sent_at = ActiveValue::set(Some(Local::now().into()));
        self.email_verification_token = ActiveValue::Set(Some(
            secret_generator.generate_token_with_length(EMAIL_VERIFICATION_TOKEN_LENGTH),
        ));
        Ok(self.update(db).await?)
    }

    /// Sets the information for a reset password request,
    /// generates a unique reset password token, and updates it in the
    /// database.
    ///
    /// This method records the timestamp when the reset password token is sent
    /// and generates a unique token for the user.
    ///
    /// # Arguments
    ///
    /// # Errors
    ///
    /// when has DB query error
    pub async fn set_forgot_password_sent(
        mut self,
        db: &DatabaseConnection,
        secret_generator: &SecretGenerator,
    ) -> ModelResult<Model> {
        const RESET_TOKEN_LENGTH: usize = 8;
        self.reset_sent_at = ActiveValue::set(Some(Local::now().into()));
        self.reset_token = ActiveValue::Set(Some(secret_generator.generate_token_with_length(RESET_TOKEN_LENGTH)));
        Ok(self.update(db).await?)
    }

    /// Records the verification time when a user verifies their
    /// email and updates it in the database.
    ///
    /// This method sets the timestamp when the user successfully verifies their
    /// email.
    ///
    /// # Errors
    ///
    /// when has DB query error
    pub async fn verified(mut self, db: &DatabaseConnection) -> ModelResult<Model> {
        self.email_verified_at = ActiveValue::set(Some(Local::now().into()));
        self.email_verification_token = ActiveValue::Set(None);
        Ok(self.update(db).await?)
    }

    /// Resets the current user password with a new password and
    /// updates it in the database.
    ///
    /// This method hashes the provided password and sets it as the new password
    /// for the user.
    ///
    /// # Errors
    ///
    /// when has DB query error or could not hashed the given password
    pub async fn reset_password(mut self, db: &DatabaseConnection, password: &str) -> ModelResult<Model> {
        self.password = ActiveValue::set(hash::hash_password(password).map_err(|e| ModelError::Any(e.into()))?);
        self.reset_token = ActiveValue::Set(None);
        self.reset_sent_at = ActiveValue::Set(None);
        Ok(self.update(db).await?)
    }
}
