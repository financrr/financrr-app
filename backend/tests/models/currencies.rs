use crate::helpers::init::init_test;
use crate::helpers::users::clean_up_user_model;
use financrr::app::App;
use financrr::models::_entities::currencies;
use insta::assert_debug_snapshot;
use loco_rs::prelude::boot_test;
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("instances");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn test_unknown_currency() {
    init_test!();

    let boot = boot_test::<App>().await.unwrap();

    let unknown_currency = currencies::Model::get_default_currency(&boot.app_context.db).await;
    assert!(unknown_currency.is_ok());

    insta::with_settings!({
        filters => clean_up_user_model()
    }, {
        assert_debug_snapshot!(unknown_currency.unwrap());
    });
}
