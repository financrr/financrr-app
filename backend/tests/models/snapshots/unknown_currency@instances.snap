---
source: tests/models/currencies.rs
assertion_line: 31
expression: unknown_currency.unwrap()
---
Model {
    id: ID,
    user_id: None,
    name: "No currency",
    symbol: "Â¤",
    iso_code: Some(
        "XXX",
    ),
    decimal_places: 0,
    created_at: DATE,
    updated_at: DATE,
}
